### PR TITLE
Fix reference urls to sections 

### PR DESCRIPTION
**Pull request by Julow**
Fix https://github.com/ocaml/odoc/issues/659

It was pointing to the wrong parent page.

### CODE REVIEW

File: src/document/url.ml [hunk 1]
```diff
@@ -232,16 +234,15 @@ module Anchor = struct
...
-     Error (Unexpected_anchor ("label " ^ anchor)) *)
+    | `Label (parent, anchor) -> (
+        let str_name = LabelName.to_string anchor in
+        (* [Identifier.LabelParent.t] contains datatypes. [`CoreType] can't
+           happen, [`Type] may not happen either but just in case, use the
+           grand-parent. *)
+        match parent with
+        | #Path.source as parent -> mk ~kind:"section" parent str_name
+        | `CoreType _ -> Error (Unexpected_anchor "core_type label parent")
+        | `Type (gp, _) -> mk ~kind:"section" gp str_name)
```
Comment (Julow) [inline src/document/url.ml:245]: `Identifier.LabelParent.t` should probably not contain types.

File: test/xref2/labels.t/run.t [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+References to the labels:
+We expect resolved references and the heading text filled in.
+
+  $ odoc_print test.odocl | jq -c '.. | .["`Reference"]? | select(.)'
+  [{"`Resolved":{"`Identifier":{"`Label":[{"`Module":[{"`Root":[{"`RootPage":"test"},"Test"]},"N"]},"B"]}}},[{"`Word":"An"},"`Space",{"`Word":"other"},"`Space",{"`Word":"conflicting"},"`Space",{"`Word":"label"}]]
+  [{"`Resolved":{"`Label":[{"`Identifier":{"`Module":[{"`Root":[{"`RootPage":"test"},"Test"]},"M"]}},"B"]}},[]]
+  [{"`Resolved":{"`Identifier":{"`Label":[{"`Root":[{"`RootPage":"test"},"Test"]},"A"]}}},[{"`Word":"First"},"`Space",{"`Word":"label"}]]
+  [{"`Resolved":{"`Identifier":{"`Label":[{"`Root":[{"`RootPage":"test"},"Test"]},"B"]}}},[{"`Word":"Floating"},"`Space",{"`Word":"label"}]]
+  [{"`Dot":[{"`Root":["M","`TUnknown"]},"C"]},[]]
```
Comment (Julow) [inline test/xref2/labels.t/run.t:25]: That's an other bug, references to labels in top-comments don't resolve.

File: src/document/url.ml (hunk 1 above)
Comment (Drup) [reply src/document/url.ml:245]: What about inlined records ?

Comment (Julow) [reply src/document/url.ml:245]: Label are section names, not fields. I don't expect a heading (sections are only defined by headings) attached to anything inside a type (classes are covered in the `Path.source` case).

File: test/xref2/dune [hunk 3]
```diff
@@ -14,8 +14,16 @@
 (cram
  (deps %{bin:odoc} %{bin:odoc_print} %{bin:compile}))
 
+(subdir
+ v407_and_above
+ (cram
+  (deps %{bin:odoc} %{bin:odoc_print} %{bin:compile})
+  (enabled_if
+   (>= %{ocaml_version} 4.07.0))))
```
Comment (Julow) [inline test/xref2/dune:22]: Before 4.07, labels in top-comments were ignored by the loader. This is probably a bug.

**Pull request merged**